### ISSUE TITLE
Migrate from aws-sdk-go to aws-sdk-go-v2

### ISSUE THREAD

**Issue by onno-vos-dev** [aws-beam/aws-codegen#103]
Opening this issue for transparency and as an announcement to the community that we will continue to support aws-elixir and aws-erlang into the future of AWS SDK v2! 🎉📢

AWS announced that AWS SDK Go (v1) will go into maintenance mode as of July 31, 2024. ⚠️ From that point on, the SDK will not receive API updates for new or existing services, or be updated to support new regions. This is problematic for aws-beam as we depend on the [aws-sdk-go](https://github.com/aws/aws-sdk-go) 🛠️ to generate the code for [aws-elixir](https://github.com/aws-beam/aws-elixir) and [aws-erlang](https://github.com/aws-beam/aws-erlang).

This issue will encapsulate the work of migrating to [aws-sdk-go-v2](https://github.com/aws/aws-sdk-go-v2) instead. 🔄

The following goals must be upheld: ✔️

* Interfaces to the clients that have integrated with either [aws-elixir](https://github.com/aws-beam/aws-elixir) and [aws-erlang](https://github.com/aws-beam/aws-erlang) MAY NOT change at all. Any changes MUST BE transparent and happen "under the hood" 🕵️
* NO new non-backwards compatible features should be included in order to make the migration transparent to the clients 🚫
* NO unexpected diff may occur in either repo. ❌ Semantically, they should produce the same code with some exceptions:
  * SuccessStatusCode is not always defined yet the new repo ensures a default of 200 whereas now we have undefined. This is broken and we can sneak in the fix in this as part of this work 🛠️
  * Some paths have changed as part of the new specification. It may be that we generated old code erroneously or misunderstood the spec. Either way, the paths will be fixed as part of the v2 API. This should in theory be fine 👍🔧
  * Docs have changed, this is expected and has no impact on compiled code. This is acceptable. 📚

**Comment by onno-vos-dev** [comment]
@jadeallenx @philss @robertoaloi @josevalim @jkakar 

This work has been ongoing inside a private repo under the `aws-beam` organisation (`aws-codegen-v2`) in order to be able to start nightly builds and compare diffs on a regular basis to ensure the above set goals are upheld. Once ready, the commit(s) will be ported over to this repo. 

The current state of this work is that we have successfully migrated and have a working version of `aws-erlang` with a minimal diff as per the goals above. We will start to look into `aws-elixir` and the diff generated there shortly to see if/what has changed there. 

aws-beam/aws-erlang#148 has been sneaked in but since that's a backwards compatible change, I figured we could wait with resolving the issue until we roll this v2 out. It's easy enough to backport in case the need arises.
